pkg/kube/dialer.go

```diff
@@ -0,0 +1,336 @@
+	addressPort := strings.Split(fqdn, ":")
+	if len(addressPort) != 2 {
+		err = ErrNoPort
+		return
+	}
+
+	port, err = strconv.Atoi(addressPort[1])
+	if err != nil {
+		return
+	}
+
+	fqdn = addressPort[0]
+	fqdn = strings.TrimSuffix(fqdn, "."+p.clusterDomain)
+	fqdn = strings.TrimSuffix(fqdn, "."+svcLabel)
+	labels := strings.Split(fqdn, ".")
+
+	// since we only dial pods we require 2 labels
+	// (assuming we're trying to dial a pod in the
+	// default namespace) or 3 (for a pod outside
+	// of the default namespace)
+	switch len(labels) {
+	case 2:
+		pod.Namespace = defaultNamespace
+	case 3:
+		pod.Namespace = labels[2]
+	default:
+		err = ErrInvalidPodFQDN
+		return
+	}
+
+	// service - labels[1]
+	pod.Name = labels[0]
+	return
+}
+
+// mutex must be held while calling this
+func (p *PodDialer) connectionForPod(pod types.NamespacedName) (httpstream.Connection, error) {
+	if conn, ok := p.connections[pod]; ok {
+		return conn, nil
+	}
+
+	transport, upgrader, err := spdy.RoundTripperFor(p.config)
+	if err != nil {
+		return nil, err
+	}
+
+	cfg := p.config
+	cfg.APIPath = "/api"
+	cfg.GroupVersion = &schema.GroupVersion{Version: "v1"}
+	cfg.NegotiatedSerializer = negotiatedSerializer
+	restClient, err := rest.RESTClientFor(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	req := restClient.Post().
+		Resource("pods").
+		Namespace(pod.Namespace).
+		Name(pod.Name).
+		SubResource("portforward")
+
+	dialer := spdy.NewDialer(upgrader, &http.Client{Transport: transport}, http.MethodPost, req.URL())
+	conn, protocol, err := dialer.Dial(portForwardProtocolV1Name)
+	if err != nil {
+		return nil, err
+	}
+	if protocol != portForwardProtocolV1Name {
+		return nil, fmt.Errorf("unable to negotiate protocol: client supports %q, server returned %q", portForwardProtocolV1Name, protocol)
+	}
+
+	p.connections[pod] = conn
+	return conn, nil
+}
+
+type conn struct {
+	httpstream.Stream
+	errorStream httpstream.Stream
+	network     string
+	remote      string
+
+	errCh chan error
+}
+
+func wrapConn(network, remote string, s, err httpstream.Stream) *conn {
+	c := &conn{
+		Stream:      s,
+		errorStream: err,
+		network:     network,
+		remote:      remote,
+		errCh:       make(chan error, 1),
+	}
+
+	go c.pollErrors()
+	return c
+}
+
+func (c *conn) pollErrors() {
+	data, err := io.ReadAll(c.errorStream)
+	if err != nil {
+		c.writeError(err)
+		return
+	}
+
+	if len(data) != 0 {
+		c.writeError(errors.New(string(data)))
+		return
+	}
+}
+
+func (c *conn) writeError(err error) {
+	select {
+	case c.errCh <- err:
+	default:
+	}
+}
+
+func (c *conn) checkError() error {
+	select {
+	case err := <-c.errCh:
+		return err
+	default:
+		return nil
+	}
+}
+
+var _ net.Conn = (*conn)(nil)
+
+func (c *conn) Read(data []byte) (int, error) {
+	if err := c.checkError(); err != nil {
+		return 0, err
+	}
+
+	n, err := c.Stream.Read(data)
+
+	// prioritize any sort of checks propagated on
+	// the error stream
+	if err := c.checkError(); err != nil {
+		return n, err
```

Comment:
nit: `errors.Join` will let you return two errors.

Reply:
Leaving this unresolved because I'm not entirely sure whether or not we'd want to join the two errors together. Thoughts?

Reply:
Let's leave this unresolved as error handling in go is a fun topic. My rule of thumb here is "If either error could contain the information I might need for debugging, I want both of them"